connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,266 @@
+
+
+### Muxer selection over Noise
+
+The libp2p Noise implementation allows the Noise handshake process to carry
+early data. [Noise-Early-Data] is carried in the second and third message of
+the XX handshake pattern as illustrated in the following message sequence chart.
+The second message carries early data in the form of a list of muxers supported
+by the responder, ordered by preference. The initiator sends its supported
+muxer list in the third message to the responder. After the Noise handshake
+process is fully done, the initiator and responder will both process the
+received eraly data and select the muxer to be used, they both iterate through
+the initiator's prefered muxer list in order, and if any muxer is also
+supported by the responder, that muxer is selected. If no mutually supported
+muxer is found, the muxer selection process MUST fall back to multistream
+-selection protocol.
+
+Example: Noise handshake between peers that have a mutually supported muxer.
+    Responder supports: ["yamux/1.0.0", "/mplex/6.7.0"]
+    Initiator supports: ["yamux/1.0.0"]
+
+    XX:
+    -> e
+    <- e, ee, s, es, ["yamux/1.0.0", "/mplex/6.7.0"] 
+    -> s, se, ["yamux/1.0.0"]
+
+    After handshake is done, both parties can arrive on the same conclusion
+    and select "yamux/1.0.0" as the muxer to use.
+
+Example: Noise handshake between peers that don't have mutually supported
+muxers.
+    Responder supports: ["/mplex/6.7.0"]
+    Initiator supports: ["yamux/1.0.0"]
+
+    XX:
+    -> e
+    <- e, ee, s, es, ["/mplex/6.7.0"]
+    -> s, se, ["yamux/1.0.0"]
+    
+    After handshaking is done, early data processing will find no mutually
+    supported muxer, and falls back to multistream-selection protocol.
+
+The muxer selection logic is run outside of the Noise handshake process. The
+format of he early data for this purpose is specified in the protobuf in the
+[Early-data-specification] section.
+
+#### Early data specification
+
+The early data message is encoded in the "protobuf2" syntax as shown in the
+following. The protobuf definition is an extension to [handshake-payload]. The
+existing byte array early data (the "data" field) will be replaced by a
+structured NoiseExtensions protobuf message. The supported muxers and selected
+muxer are populated in the "stream_muxers" field. The details of the early
+data message can be find in [Noise-handshake-payload]
+
+The muxers are ordered by preference, with the most prefered muxer at the
+beginning.
+
+```protobuf
+syntax = "proto2";
+
+message NoiseExtensions {
+    repeated bytes webtransport_certhashes = 1;
+    repeated string stream_muxers = 2; 
+}
+
+message NoiseHandshakePayload {
+  optional bytes identity_key = 1;
+  optional bytes identity_sig = 2;
+  optional NoiseExtensions extensions = 4;
+}
+```
+
+## Cross version support
+
+The improved muxer selection approach MUST be interoperable with pervious
+libp2p versions which do not support this improved approach.
+
+### TLS case
```

Comment:
Explain how backwards compatibility works in the TLS section of this document?

Reply:
done